osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -20,51 +20,165 @@ public class Speed : OsuStrainSkill
+        {
+            if (current.BaseObject is Spinner)
+                return 0;
+
+            var osuCurrObj = (OsuDifficultyHitObject)current;
+
+            return speedBonus / osuCurrObj.StrainTime;
+        }
+
+        private double movementStrainOf(DifficultyHitObject current, double speedBonus)
+        {
+            if (current.BaseObject is Spinner)
+                return 0;
+
+            var osuCurrObj = (OsuDifficultyHitObject)current;
+
+            double distance = Math.Min(single_spacing_threshold, osuCurrObj.TravelDistance + osuCurrObj.JumpDistance);
 
             double angleBonus = 1.0;
 
-            if (osuCurrent.Angle != null && osuCurrent.Angle.Value < angle_bonus_begin)
+            if (osuCurrObj.Angle != null)
             {
-                angleBonus = 1 + Math.Pow(Math.Sin(1.5 * (angle_bonus_begin - osuCurrent.Angle.Value)), 2) / 3.57;
+                double angle = osuCurrObj.Angle.Value;
+
+                if (angle < pi_over_2)
+                    angleBonus = 1.25;
+                else if (angle < angle_bonus_begin)
+                    angleBonus = 1 + Math.Pow(Math.Sin(1.5 * (angle_bonus_begin - angle)), 2) / 4;
 
-                if (osuCurrent.Angle.Value < pi_over_2)
-                {
-                    angleBonus = 1.28;
-                    if (distance < 90 && osuCurrent.Angle.Value < pi_over_4)
-                        angleBonus += (1 - angleBonus) * Math.Min((90 - distance) / 10, 1);
-                    else if (distance < 90)
-                        angleBonus += (1 - angleBonus) * Math.Min((90 - distance) / 10, 1) * Math.Sin((pi_over_2 - osuCurrent.Angle.Value) / pi_over_4);
-                }
             }
 
-            return (1 + (speedBonus - 1) * 0.75) * angleBonus * (0.95 + speedBonus * Math.Pow(distance / single_spacing_threshold, 3.5)) / osuCurrent.StrainTime;
+            return (angleBonus * speedBonus * Math.Pow(distance / single_spacing_threshold, 3.5)) / osuCurrObj.StrainTime;
+        }
+
+        private double strainDecay(double ms) => Math.Pow(strainDecayBase, ms / 1000);
+
+        protected override double CalculateInitialStrain(double time) => (currentMovementStrain + currentTapStrain * currentRhythm) * strainDecay(time - Previous[0].StartTime);
+
+        protected override double StrainValueAt(DifficultyHitObject current)
+        {
+            double speedBonus = 1.0;
+            double deltaTime = Math.Max(max_speed_bonus, current.DeltaTime);
+
+            if (deltaTime < min_speed_bonus)
+                speedBonus = 1 + 0.75 * Math.Pow((min_speed_bonus - deltaTime) / speed_balancing_factor, 2);
+
+            currentRhythm = calculateRhythmBonus(current);
+
+            currentTapStrain *= strainDecay(current.DeltaTime);
+            currentTapStrain += tapStrainOf(current, speedBonus) * skillMultiplier;
+
+            currentMovementStrain *= strainDecay(current.DeltaTime);
```

Comment:
Strain decay could be computed once as entirely dependent on `DeltaTime`.

Reply:
Done.